### PR TITLE
Fix openshift controller manager shutdown sequence

### PR DESCRIPTION
**Pull request by mfojtik**
Wire stop channel and signals to openshift controller manager to shutdown properly and remove SDNOTIFY.

/cc @deads2k 
/cc @sttts

### CODE REVIEW

File: pkg/cmd/openshift-kube-apiserver/cmd.go [hunk 1]
```diff
@@ -70,6 +70,11 @@ func NewOpenShiftKubeAPIServerServerCommand(name, basename string, out, errout i
 				}
 				klog.Fatal(err)
 			}
+
+			// When no error is returned, always return with zero exit code.
+			// This is here to make sure the container that run apiserver won't get accidentally restarted
+			// when the pod runs with restart on failure.
+			os.Exit(0)
```
Comment (sttts) [inline pkg/cmd/openshift-kube-apiserver/cmd.go]: isn't this implicit?

Comment (mfojtik) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: i think so, removed (but left the comment)

File: pkg/cmd/openshift-kube-apiserver/cmd.go [hunk 2]
```diff
@@ -70,14 +70,21 @@ func NewOpenShiftKubeAPIServerServerCommand(name, basename string, out, errout i
...
 		},
 	}
 
 	flags := cmd.Flags()
 	// This command only supports reading from config
 	flags.StringVar(&options.ConfigFile, "config", "", "Location of the master configuration file to run from.")
-	cmd.MarkFlagFilename("config", "yaml", "yml")
-	cmd.MarkFlagRequired("config")
+	if err := cmd.MarkFlagFilename("config", "yaml", "yml"); err != nil {
+		klog.Warningf("Error setting config flag: %v", err)
```
Comment (sttts) [inline pkg/cmd/openshift-kube-apiserver/cmd.go]: what does it mean to get an error here?

File: pkg/cmd/openshift-kube-apiserver/cmd.go [hunk 3]
```diff
@@ -70,14 +70,21 @@ func NewOpenShiftKubeAPIServerServerCommand(name, basename string, out, errout i
...
+	}
+	if err := cmd.MarkFlagRequired("config"); err != nil {
+		klog.Warningf("Error setting config flag required: %v", err)
```
Comment (sttts) [inline pkg/cmd/openshift-kube-apiserver/cmd.go]: what does it mean to get an error here?

File: pkg/cmd/openshift-kube-apiserver/cmd.go (hunk 2 above)
Comment (mfojtik) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: for example the flag does not exists (or is removed). in that case this throws error. goland is unhappy when you don't catch these and I think in general we should catch all errors ;-)

File: pkg/cmd/openshift-kube-apiserver/cmd.go (hunk 3 above)
Comment (mfojtik) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: same answer as above

File: pkg/cmd/openshift-kube-apiserver/cmd.go (hunk 2 above)
Comment (sttts) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: was wondering whether a panic makes more sense here.

Comment (mfojtik) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: i don't think this is panic worthy, @deads2k ?

File: pkg/cmd/openshift-controller-manager/controller_manager.go [hunk 4]
```diff
@@ -93,7 +93,16 @@ func RunOpenShiftControllerManager(config *openshiftcontrolplanev1.OpenShiftCont
...
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	go func() {
+		select {
+		case <-stopCh:
+			cancel()
+		}
+	}()
+
+	go leaderelection.RunOrDie(ctx,
```
Comment (sttts) [inline pkg/cmd/openshift-controller-manager/controller_manager.go]: wasn't there a new RunWithContext?

Comment (sttts) [reply pkg/cmd/openshift-controller-manager/controller_manager.go]: oh, that's it. I thought to remember more contexts passed everywhere.

File: pkg/cmd/openshift-controller-manager/cmd.go [hunk 5]
```diff
@@ -162,5 +165,11 @@ func (o *OpenShiftControllerManager) RunControllerManager() error {
 		return err
 	}
 
-	return RunOpenShiftControllerManager(config, clientConfig)
+	if err := RunOpenShiftControllerManager(config, clientConfig, stopCh); err != nil {
+		return err
+	}
+
+	<-stopCh
```
Comment (sttts) [inline pkg/cmd/openshift-controller-manager/cmd.go]: RunOpenShiftControllerManager blocks on stopCh already.

File: pkg/cmd/openshift-controller-manager/cmd.go [hunk 6]
```diff
@@ -74,8 +77,12 @@ func NewOpenShiftControllerManagerCommand(name, basename string, out, errout io.
 	flags := cmd.Flags()
 	// This command only supports reading from config
 	flags.StringVar(&options.ConfigFilePath, "config", options.ConfigFilePath, "Location of the master configuration file to run from.")
-	cmd.MarkFlagFilename("config", "yaml", "yml")
-	cmd.MarkFlagRequired("config")
+	if err := cmd.MarkFlagFilename("config", "yaml", "yml"); err != nil {
```
Comment (soltysh) [inline pkg/cmd/openshift-controller-manager/cmd.go]: I think in these two cases you can ignore the error, we never verify it.

File: pkg/cmd/openshift-kube-apiserver/cmd.go (hunk 2 above)
Comment (soltysh) [reply pkg/cmd/openshift-kube-apiserver/cmd.go]: We don't worry about that anywhere, it's safe to ignore that one error. Not even warning worthy.

File: pkg/cmd/openshift-controller-manager/cmd.go (hunk 5 above)
Comment (mfojtik) [reply pkg/cmd/openshift-controller-manager/cmd.go]: yeah, figured, fixed.

Comment (sttts) [reply pkg/cmd/openshift-controller-manager/cmd.go]: why so complicated? do it the other way around. Unneccessary concurrency is not nice.

**Pull request closed without merging**